### DISCUSSION TITLE
Change LAF (look-and-feel) through plugin

### DISCUSSION THREAD

**Discussion by MatthiasDM** [nextgenhealthcare/connect#5917] [Feature/Enhancement Ideas]
Hi,
Would it be possible to change the look and feel of the interface through a plugin? Would be nice if the interface looked somewhat more modern.

**Comment by jonbartels** [comment]
The look and feel is not pluggable. I am trying to find notes and references about the look and feel

https://github.com/nextgenhealthcare/connect/issues/3935 is the closest Mirth has ever come to supporting themes.

You should also look at threads from [this forum search](https://forums.mirthproject.io/search?q=theme&searchJSON=%7B%22keywords%22%3A%22theme%22%7D) . There might be more details in the old forum posts than I can remember!

From my memory, a common request from my employers and customers is to change the logo. From what I recall even such a simple change requires recompiling code and also recompiling the client-side JARs. The logos are packaged with the JARs. 

The Mirth theme is handled by [JGoodies](https://www.jgoodies.com/). Searching the code for `com.jgoodies` should find references about how it is used. In theory, JGoodies can support pluggable themes, however Mirth does not use JGoodies in a way that is pluggable.

Mirth specifically uses [JGoodies Looks](https://www.jgoodies.com/freeware/libraries/looks/). It is not clear to me if Looks is still free under a BSD license or not.

I think the starting point to make the themes swappable or pluggable is:

1. Find out what JGoodies supports and how it supports themes
2. I think the themes would be first injected in this code: https://github.com/nextgenhealthcare/connect/blob/33977cb2a76828203957a39af804e3f3f7600994/client/src/com/mirth/connect/client/ui/Mirth.java#L166
3. There appear to be some tweaks or customizations to what JGoodies does based on a search for `com.jgoodies`
4. Mirth uses JGoodies from the `looks-2.3.1.jar` dependency. Is 2.3.1 a recent release? Do newer versions of JGoodies offer better features for swapping themes?

I hope that my memory is accurate. If you start hacking on this problem please post about it!!!

**Reply by MatthiasDM** [reply]
Will do, thank you for your detailed response!

**Comment by MatthiasDM** [comment]
I've had moderate success implementing a new LAF (FlatLAF) but due to the many custom components it has proven to be difficult to change all UI-elements. I'll try to change to logo and give some feedback.
